lib/connection_controller.go

```diff
@@ -231,6 +252,84 @@ func (cc *ConnectionController) refreshConnectIps() {
 	}
 }
 
+// ###########################
+// ## Validator Connections
+// ###########################
+
+// refreshValidatorIndex re-indexes validators based on the activeValidatorsMap. It is called periodically by the
+// validator connector.
+func (cc *ConnectionController) refreshValidatorIndex(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, *ValidatorEntry]) {
+	// De-index inactive validators. We skip any checks regarding RemoteNodes connection status, nor do we verify whether
+	// de-indexing the validator would result in an excess number of outbound/inbound connections. Any excess connections
+	// will be cleaned up by the peer connector.
+	validatorRemoteNodeMap := cc.rnManager.GetValidatorIndex().Copy()
+	for pk, rn := range validatorRemoteNodeMap {
+		// If the validator is no longer active, de-index it.
+		if _, ok := activeValidatorsMap.Get(pk); !ok {
+			cc.rnManager.SetNonValidator(rn)
+			cc.rnManager.UnsetValidator(rn)
```

Comment:
Would it make sense to disconnect from these validators instead? Ex: what happens if have the connectIps flag as non-empty, a validator leaves the validator set: we would end up with an extra non-validator connection that we don't intend to maintain

Reply:
It's possible this inactive validator will be disconnected later by the NonValidator connector, which makes decisions about expunging from the NonValidator index. As I was thinking about the organization of the NetworkManager, this design seemed to result in a cleaner separation of concerns.

As for the example you mentioned, the connectIps RemoteNodes (persistent RemoteNodes) get "special treatment" in that we'll never intentionally disconnect from them. Persistent peers are also not included in the count for targetOutbound/InboundRemoteNodes. The only way for us to drop persistent RemoteNodes will be them closing the connection with us.